### ISSUE TITLE
MMDevice.Dispose() hangs indefinitely

### ISSUE THREAD

**Issue by mkahvi** [naudio/NAudio#449]
Any instance of `MMDevice.Dispose()` is causing a thread to hang indefinitely.

I feel like it should throw an error of some sort instead in case it's an issue of disposing it in wrong thread (noticed there were some other reported issues with doing stuff with MMDevice outside of the thread it was created).

All I've done with the MMDevice is to subscribe to `MMDevice.AudioEndpointVolume.OnVolumeNotification` events, in case that matters.

**Comment by mkahvi** [comment]
I think the above is related to `UnregisterEndpointNotificationCallback` hanging indefinitely as well, since removing the MMDevice.Dispose() causes it to work as expected.

**Comment by mkahvi** [comment]
Testing seems to point out that this is caused by calling `Dispose` on some other thread than the one `MMDevice` was created on. Either it should throw or warn about the need for same thread context.

**Comment by markheath** [comment]
Yes, there are some nasty gotchas with threading on COM objects. I know other similar libraries improve the situation with post-compile steps to use the CallI IL instruction, which is something I'd like to do for NAudio in the future.

**Comment by markheath** [comment]
I am currently using AI to help catch up on the backlog of open issues, and so updating with the latest state before closing...

You diagnosed this correctly back in 2019: the hang was a classic COM apartment-threading problem. NAudio's Core Audio layer used thread-affine runtime-callable wrappers, so calling `Dispose()` (and therefore `UnregisterControlChangeNotify`) from a thread in a different apartment than the one that created the `MMDevice` forced a cross-apartment marshal that could block forever if the original (typically STA UI) thread wasn't available or pumping messages.

The good news: the entire `NAudio.Wasapi` Core Audio interop layer has since been rewritten (NAudio 3, current `main`) to use .NET source-generated COM with `ComWrappers` instead of classic RCWs. Among other things this was done specifically to remove the "created on a WinForms/WPF STA thread, used from an MTA audio thread" failure mode, so the deadlock mechanism you hit no longer applies — `MMDevice`/`AudioEndpointVolume` disposal no longer marshals back to a creating apartment.

One thing still worth being aware of: if you capture the `AudioEndpointVolume` on a UI thread, `OnVolumeNotification` events are posted back to that `SynchronizationContext`, so a blocked UI thread can still stall notification delivery (not disposal). Disposing the `MMDevice` deterministically when you are done unsubscribing is the recommended pattern.

I'm closing this issue now, but of course if you find you still have problems, let me know and we can reopen and investigate
